src/emulator/ui/server.ts

```diff
@@ -0,0 +1,118 @@
+/**
+ * Copyright 2024 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { createServer } from 'http';
+import type { ListenOptions } from 'net';
+import * as path from 'path';
+
+import * as express from "express";
+import fetch from 'node-fetch';
+import { AnalyticsSession } from "../../track";
+import { ExperimentName } from "../../experiments";
+import * as http from "http";
+
+/*
+  This file defines Node.js server-side logic for the Emulator UI.
+
+  It is a facsimile class of the local server in firebase-tools-ui/server.ts. For that reason, 
+  values that were previously environment variables are passed in rather than fetched/derived 
+  in the class itself.
+*/
+export async function createApp(
+  zipDirPath: string,
+  projectId: string,
+  hubHost: string,
+  emulatorGaSession: AnalyticsSession | undefined,
+  listenOptions: ListenOptions[],
+  experiments: Array<ExperimentName>): Promise<http.Server> {
+
+  const app = express();
+  // Exposes the host and port of various emulators to facilitate accessing
+  // them using client SDKs. For features that involve multiple emulators or
+  // hard to accomplish using client SDKs, consider adding an API below.
+  app.get(
+    '/api/config',
+    jsonHandler(async () => {
+      const hubDiscoveryUrl = new URL(`http://${hubHost}/emulators`);
+      const emulatorsRes = await fetch(hubDiscoveryUrl.toString());
+      const emulators = (await emulatorsRes.json()) as any;
+
+      const json = { projectId, experiments: [], ...emulators };
+
+      // Googlers: see go/firebase-emulator-ui-usage-collection-design?pli=1#heading=h.jwz7lj6r67z8
+      // for more detail
+      if (emulatorGaSession) {
+        json.analytics = emulatorGaSession;
+      }
+
+      // pick up any experiments enabled with `firebase experiment:enable`
+      if (experiments) {
+        json.experiments = experiments;
+      }
+
+      return json;
+    })
+  );
+
+  const webDir = path.join(zipDirPath, 'client');
+  app.use(express.static(webDir));
+  // Required for the router to work properly.
+  app.get('*', function (_, res) {
+    res.sendFile(path.join(webDir, 'index.html'));
+  });
+
+  if (listenOptions.length == 0) {
+    console.error(`Failed to start UI server, listenOptions empty`);
+    process.exit(1);
+  }
+  var server = null;
+  for (const opts of listenOptions) {
+    server = createServer(app).listen(opts);
```

Comment:
This will only return the last server to `ui.ts` and only one destroyer will be created in `ui.ts`. This leads to servers not properly shutting down. `ExpressBasedEmulator` handles all the listening etc. correctly -- let's just inherit the class and reuse the logic there. No need to duplicate the work (Line 77-97 here).

I think the `createApp` function in a separate file is hindering our ability to reuse code in this case. Consider making it a method in class EmulatorUI instead. e.g. `override createExpressApp() { ... }`

Reply:
you're right - I want to keep this in line with the server.ts in firebase-tools-ui but maybe that's a painful goal to have - this should work better, though still flawed

Reply:
I don't think we need to keep this in line with server.ts in firebase-tools-ui. After this PR gets merged, the file in the UI can be rewritten to serve development only. That file may get removed or changed so much that it no longer resembles anything here.